### PR TITLE
[Feature] Add support to upload files

### PR DESCRIPTION
**Pull request by SferaDev**
- https://app.clickup.com/t/d31bur

### CODE REVIEW

File: src/api/files.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        const response = this.d2Api.apiConnection
+            .request<PartialSaveResponse>({
+                method: "post",
+                url: "/fileResources",
+                data: formData,
+                dataType: "raw",
+            })
+            .getData()
+            .then(data => {
```
Comment (tokland) [inline src/api/files.ts:44]: Can you check if `.flatMap(res => ` works here?

File: src/data/AxiosHttpClientRepository.ts [hunk 2]
```diff
@@ -1,14 +1,14 @@
-import MockAdapter from "axios-mock-adapter";
 import axios, { AxiosInstance } from "axios";
+import MockAdapter from "axios-mock-adapter";
 import qs from "qs";
-
+import { CancelableResponse } from "../repositories/CancelableResponse";
 import {
+    ConstructorOptions,
+    getBody,
```
Comment (tokland) [inline src/data/AxiosHttpClientRepository.ts:7]: Unused. No body processing for axios repo?

Comment (SferaDev) [reply src/data/AxiosHttpClientRepository.ts:7]: Nope, axios auto detects the body and processes it internally ie form data

File: src/api/files.ts (hunk 1 above)
Comment (SferaDev) [reply src/api/files.ts:44]: You mean something like this?

```ts
return this.d2Api.apiConnection
    .request<PartialSaveResponse>({
        method: "post",
        url: "/fileResources",
        data: formData,
        dataType: "raw",
    })
    .flatMap(({ data }) => {
        const fileResourceId =
            data.response && data.response.fileResource
            ? data.response.fileResource.id
            : undefined;

        const document = { id, name, url: fileResourceId };

        return this.d2Api
            .post<MetadataResponse>("/metadata", {}, { documents: [document] })
            .map(({ data }) => ({ id, response: data }));
    });
}
```

Comment (tokland) [reply src/api/files.ts:44]: That's it!

Comment (SferaDev) [reply src/api/files.ts]: Seems to work just fine, will publish and re-test without ``yarn link``

**Review by tokland — APPROVED** [review]

**Pull request merged**